libs/constrained-generators/src/Constrained/Base.hs

```diff
@@ -1532,29 +1623,27 @@ stepPlan (SolverPlan (SolverStage x ps spec : pl) gr) = do
           )
           (spec <> spec')
       )
-  let env = singletonEnv x val
-  pure (env, backPropagation $ SolverPlan (substStage env <$> pl) (deleteNode (Name x) gr))
+  let env1 = extendEnv x val env
+  pure (env1, backPropagation $ SolverPlan (substStage env1 <$> pl) (deleteNode (Name x) gr))
 
 substStage :: Env -> SolverStage fn -> SolverStage fn
 substStage env (SolverStage y ps spec) = normalizeSolverStage $ SolverStage y (substPred env <$> ps) spec
 
 -- | Generate a satisfying `Env` for a `p : Pred fn`. The `Env` contains values for
 -- all the free variables in `flattenPred p`.
-genFromPreds :: (MonadGenError m, BaseUniverse fn) => Pred fn -> GenT m Env
+genFromPreds :: forall fn m. (MonadGenError m, BaseUniverse fn) => Env -> Pred fn -> GenT m Env
 -- TODO: remove this once optimisePred does a proper fixpoint computation
-genFromPreds (optimisePred . optimisePred -> preds) = explain1 (show $ "genFromPreds fails\nPreds are:" /> pretty preds) $ do
-  -- NOTE: this is just lazy enough that the work of flattening, computing dependencies,
-  -- and linearizing is memoized in properties that use `genFromPreds`.
-  plan <- runGE $ prepareLinearization preds
-  explain1 (show $ pretty plan) $
-    go mempty plan
+genFromPreds env0 preds0 = do
```

Comment:
I don't like losing the explanatory comments here. Also, it would be nice if we could keep the old, simpler, interface for the function.

Reply:
Maybe I am not understanding this, but if we print the plan, and then later print each step (while stepping the plan) everything appears twice in the error message. If we only print the steps, and in each successful step, just print the value generated, then the steps appear line this
```
step x <- 5
step ys <- [3,4,5]
step z <- Map.fromList [("a",3),("b",9)]
step w fails where spec = (SetSpec must elem size)
... (bunch of error messages)
```
It is much more compact, and tells the important information
1) The order of solving the variables
2) The actual value computed
3) The actual spec that failed
4) all the explanations about why

I can leave the more verbose things as comments, and if you don't like it, we can revert back. But the way it is now is very helpfull.